daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -599,6 +599,7 @@ func (devices *DeviceSet) cleanupDeletedDevices() error {
 
 	// If there are no deleted devices, there is nothing to do.
 	if devices.nrDeletedDevices == 0 {
+		devices.Unlock()
```

Comment:
@lk4d4 I agree that this looks odd. In general reason is that when device deletion is happning, we can try deletion/deactivation in a loop. And that that time we don't want to block rest of the device operations in parallel. So we drop the inner devices lock while continue to hold per device lock. But that also seems to require that per device lock be acquired before devices lock and all these complications arise. 

I hope there is a better way to design this. I will keep this in mind.